numerical aquifers

GitPaean opened the pull request:
As requested, PR https://github.com/OPM/opm-common/pull/2097 was requested to be split for easier review. 

This is the most trivial part.  It just parses and generates the numerical aquifer information.  I am not familiar with the serialization part, especially when pointers are involved. Some help is needed here. Thanks in advance. 

There will be another one after this. 

https://github.com/OPM/opm-common/pull/2208 is also one needs to be addressed. 

Hopefully, with splitting up we can get proper implementation of numerical aquifer in.  Thanks.

GitPaean commented:
We have not applied numerical aquifers related information for anything. Basically, it should not change anything in the master branch running.

GitPaean commented:
jenkins build this please

GitPaean commented on opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquifers.hpp:
```diff
@@ -0,0 +1,65 @@
+
+    private:
+        std::unordered_map <size_t, SingleNumericalAquifer> aquifers_;
+        std::unordered_map<size_t, const NumericalAquiferCell*> aquifer_cells_;
```
This is for convenience of later usage.  Maybe should be moved out of this PR.

GitPaean commented on opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferConnection.hpp:
```diff
@@ -0,0 +1,73 @@
+    class Deck;
+    class DeckRecord;
+
+    struct NumericalAquiferConnection {
```
There is some parts in common with the connection of the two analytical aquifer models (the way to determine whether the face can be connection face), and also some parts are pretty different.  Some refactoring might be done to remove some small duplication code (basically the code of determination of whether a face can be a connection).

GitPaean replied:
Preferably, we can do the refactoring in a separate PR. Because that will be involved in refactoring of the existing aquifer models.

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferConnection.cpp:
```diff
@@ -0,0 +1,133 @@
+    {
+        std::map<size_t, std::map<size_t, NumericalAquiferConnection>> connections;
+        using AQUCON=ParserKeywords::AQUCON;
+        if ( !deck.hasKeyword<AQUCON>() ) return connections;
```
I prefer:
```
if (!deck.hasKeyword<>)
    return {}

std::map<size_t, std::map<size_t, NumericalAquiferConnection>> connections;
```

joakim-hove commented:
What is here looks fine - but .... I am not familiar with the downstream usage, but I get the feeling that what is needed from downstream is:

1. A container - e.g. `NumericalAquifers`
2. A way to get all the aquifer cells.

If this is correct I suggest that the `SingleNumericalAquifer` is kept internally in opm-common and not installed. I don't fully understand the details - but I don't think you really to store the map of pointer to aquifercells - I would assume that could be created on request?

joakim-hove commented on opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferCell.hpp:
```diff
@@ -0,0 +1,69 @@
+        double length;
+        double porosity;
+        double permeability;
+        double depth; // by default the grid block depth will be used
```
No need to add the "by default ...." comments - this is transparently handled in the cpp implementation and the reader does not need to know.

joakim-hove commented on opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferCell.hpp:
```diff
@@ -0,0 +1,69 @@
+    struct NumericalAquiferCell {
+        NumericalAquiferCell(const DeckRecord&, const EclipseGrid&, const FieldPropsManager&);
+        NumericalAquiferCell() = default;
+        size_t aquifer_id; // aquifer id
```
These comments are only noise - please remove them!

joakim-hove commented on opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquifers.hpp:
```diff
@@ -0,0 +1,65 @@
+        }
+
+    private:
+        std::unordered_map <size_t, SingleNumericalAquifer> aquifers_;
```
Please - no trailing `_` in this part of town; either just plain `aquifers` or `m_aquifers`

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferCell.cpp:
```diff
@@ -0,0 +1,105 @@
+            this->sattable = satnum[active_index];
+        }
+
+        this->transmissibility = 2. * this->permeability * this->area / this->length;
```
I guess the transmissibility could be a function like `cellVolume()`?

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferConnection.cpp:
```diff
@@ -0,0 +1,133 @@
+        std::vector<NumericalAquiferConnection> cons;
+
+        const size_t i1 = record.getItem<AQUCON::I1>().get<int>(0) - 1;
+        const size_t j1 = record.getItem<AQUCON::J1>().get<int>(0) -1;
```
That alignment tickles my OCD

GitPaean replied:
if you are referring to `-1;`, I fixed it just now.

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferConnection.cpp:
```diff
@@ -0,0 +1,133 @@
+        const size_t j2 = record.getItem<AQUCON::J2>().get<int>(0) -1;
+        const size_t k2 = record.getItem<AQUCON::K2>().get<int>(0) - 1;
+
+        const std::string str_allow_internal_cells = record.getItem<AQUCON::ALLOW_INTERNAL_CELLS>().getTrimmedString(0);
```
No reason to go via the temporary local string variable?

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquiferConnection.cpp:
```diff
@@ -0,0 +1,133 @@
+                    if (allow_internal_cells ||
+                        !AquiferHelpers::neighborCellInsideReservoirAndActive(grid, i, j, k, face_dir)) {
+                        const size_t global_index = grid.getGlobalIndex(i, j, k);
+                        cons.emplace_back(NumericalAquiferConnection{i, j, k, global_index, allow_internal_cells, record});
```
When you use `emplace_back()` the arguments to `emplace_back()` are forwarded to the constructor - i.e. this should be written:
```
cons.emplace_back( i, j, k, global_index, allow_internal_cells, record );
```

joakim-hove commented on src/opm/parser/eclipse/EclipseState/Aquifer/NumericalAquifer/NumericalAquifers.cpp:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    const SingleNumericalAquifer& NumericalAquifers::getAquifer(const size_t aquifer_id) const {
+        if ( !this->hasAquifer(aquifer_id)) {
```
The "check-and-return-if-found" function is most efficiently implemented with iterators and the `find()` method.

GitPaean commented:
> . I am not familiar with the downstream usage, but I get the feeling that what is needed from downstream is:

>. A container - e.g. NumericalAquifers
>. A way to get all the aquifer cells.
> If this is correct I suggest that the SingleNumericalAquifer is kept internally in opm-common and not installed. 

Form the simulator side, you can refer to https://github.com/OPM/opm-simulators/blob/22ed97868df3564c8eb552d4f7decdc34c7dc817/opm/simulators/aquifers/AquiferNumerical.hpp . 
And you are right, with that implementation. The main information is the aquifer cells. (the connections are becoming part of transmissbility). 

>  I don't fully understand the details - but I don't think you really to store the map of pointer to aquifercells - I would assume that could be created on request?

Yes, it is possible. We only need it in a few places. 

> Update: Having looked a bit more on this it looks like the whole SingleNumericalAquifer class can be dropped it in its' entirety?

How so? For sure, we can achieve the same thing with std::vector or std::map, but I do not think it will make the things easier.  There will be more actions related to numerical aquifers in the coming PR, I think it is good that the in the code an aquifer is also taking care of its own information.

GitPaean commented:
Thanks for all the comments except making `SingleNumericalAquifer` internal.  If you can give more specific instructions, I will do it.  It looks like the member of `NumericalAquifers` need this class to be complete, and can not use forward declaration there.

GitPaean commented:
jenkins build this please

GitPaean commented:
> Thanks for all the comments except making SingleNumericalAquifer internal. 

O~~. I want to say thanks for all the comments.   I have addressed all the comments except making SingleNumericalAquifer internal.
